File: android/java/IndoorLocationExample/app/src/main/java/com/gipstech/indoorlocationexample/MainActivity.java
package com.gipstech.indoorlocationexample;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.widget.TextView;

import com.gipstech.*;

public class MainActivity extends AppCompatActivity {
    final static String DEV_KEY = ""; // PUT YOUR DEV KEY HERE
    final static String BUILDING_ID = ""; // PUT THE ID OF THE BUILDING HERE

    TextView textView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        textView = findViewById(R.id.mainTextView);
        textView.setMovementMethod(new ScrollingMovementMethod());

        String[] permissions = new String[]{
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.BLUETOOTH
        };

        GiPStech.init(this, DEV_KEY);

        if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
                checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
                checkSelfPermission(Manifest.permission.BLUETOOTH) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(permissions, 0);
        } else {
            startLocation();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (!allGranted(grantResults)) {
            textView.append("Please grant all permissions to start location");
            return;
        }

        startLocation();
    }

    boolean allGranted(int[] grantResults) {
        if (grantResults.length == 0) {
            return false;
        }

        for (int val : grantResults) {
            if (val != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }

        return true;
    }

    void performFastCalibration() {
        GiPStech.getCalibrationManager().beginCalibration(CalibrationManager.CALIBRATION_TYPE_FAST, new ProgressCallback<Void>() {
            @Override
            public void onResult(Void result) {
                textView.append("Calibration completed\n");
                GiPStech.getIndoorLocalizer().registerListener(listener);
            }

            @Override
            public void onProgress(int progress) {
                textView.append("Calibration progress: " + progress + "%\n");
                if (progress >= 25) {
                    GiPStech.getCalibrationManager().endCalibration();
                }
            }

            @Override
            public void onError(GiPStechError error) {
                textView.append(error.getMessage() + "\n");
            }
        });
    }

    void startLocation() {
        GiPStech.getIndoorLocalizer().selectBuilding(BUILDING_ID, new ProgressCallback<Void>() {
            @Override
            public void onResult(Void result) {
                textView.append("Selection completed\n");
                GiPStech.getIndoorLocalizer().registerListener(listener);
            }

            @Override
            public void onProgress(int progress) {
                textView.append("Selection progress: " + progress + "%\n");
            }

            @Override
            public void onError(GiPStechError error) {
                textView.append(error.getMessage() + "\n");
            }
        });
    }

    IndoorLocalizer.Listener listener = new IndoorLocalizer.Listener() {
        @Override
        public void onRegistrationUpdate(int percentage) {
            textView.append("Registration progress: " + percentage + "%\n");
        }

        @Override
        public void onRegistrationComplete() {
            textView.append("Registration completed\n");
        }

        @Override
        public void onLocationUpdate(IndoorLocation location) {
            textView.append("Location: " + location.getLongitude() + " " + location.getLatitude() + "\n");
        }

        @Override
        public boolean onFloorChange(Floor floor) {
            textView.append("Floor: " + floor.getName() + "\n");
            return true;
        }

        @Override
        public void onAttitudeUpdate(Attitude attitude) {
            textView.append("Heading: " + attitude.getHeading() + "\n");
        }

        @Override
        public void onError(GiPStechError error) {
            textView.append(error.getMessage() + "\n");

            if (GiPStechError.FAST_CALIBRATION_REQUIRED == error.getCode()) {
                performFastCalibration();
            }
        }
    };
}

File: android/kotlin/IndoorLocationExample/app/src/main/java/com/gipstech/indoorlocationexample/MainActivity.kt
package com.gipstech.indoorlocationexample

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.method.ScrollingMovementMethod
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

import com.gipstech.*

class MainActivity : AppCompatActivity() {
    val DEV_KEY = "" // PUT YOUR DEV KEY HERE
    val BUILDING_ID = "" // PUT THE ID OF THE BUILDING HERE

    var textView: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        textView = findViewById(R.id.mainTextView)
        textView?.movementMethod = ScrollingMovementMethod()

        val permissions = arrayOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.BLUETOOTH)

        GiPStech.init(this, DEV_KEY)

        if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
            checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED ||
            checkSelfPermission(Manifest.permission.BLUETOOTH) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(permissions, 0)
        } else {
            startLocation()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (!allGranted(grantResults)) {
            textView?.append("Please grant all permissions to start location")
            return
        }

        startLocation()
    }

    fun allGranted(grantResults: IntArray): Boolean {
        if (grantResults.isEmpty()) {
            return false
        }
        for (res in grantResults) {
            if (res != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }
        return true
    }

    fun performFastCalibration() {
        GiPStech.getCalibrationManager().beginCalibration(CalibrationManager.CALIBRATION_TYPE_FAST, object: ProgressCallback<Void?> {
            override fun onResult(result: Void?) {
                textView?.append("Calibration completed\n")
                GiPStech.getIndoorLocalizer().registerListener(listener)
            }

            override fun onProgress(progress: Int) {
                textView?.append("Calibration progress: " + progress + "%\n")

                if (progress >= 25) {
                    GiPStech.getCalibrationManager().endCalibration()
                }
            }

            override fun onError(error: GiPStechError) {
                textView?.append(error.message + "\n")
            }
        })
    }

    fun startLocation() {
        GiPStech.getIndoorLocalizer().selectBuilding(BUILDING_ID, object: ProgressCallback<Void?> {
            override fun onResult(result: Void?) {
                textView?.append("Selection completed\n")
                GiPStech.getIndoorLocalizer().registerListener(listener)
            }

            override fun onProgress(progress: Int) {
                textView?.append("Selection progress: " + progress + "%\n")
            }

            override fun onError(error: GiPStechError) {
                textView?.append(error.message + "\n")
            }
        })
    }

    val listener = object: IndoorLocalizer.Listener {
        override fun onRegistrationComplete() {
            textView?.append("Registration completed\n")
        }

        override fun onRegistrationUpdate(progress: Int) {
            textView?.append("Registration progress: " + progress + "%\n")
        }

        override fun onLocationUpdate(location: IndoorLocation) {
            textView?.append("Location: " + location.longitude + " " + location.latitude + "\n")
        }

        override fun onFloorChange(floor: Floor): Boolean {
            textView?.append("Floor: " + floor.name + "\n")
            return true
        }

        override fun onAttitudeUpdate(attitude: Attitude) {
            textView?.append("Heading: " + attitude.heading + "\n")
        }

        override fun onError(error: GiPStechError) {
            textView?.append(error.message + "\n")

            if (GiPStechError.FAST_CALIBRATION_REQUIRED == error.code) {
                performFastCalibration()
            }
        }
    }
}